april_tag_updates/docking_script.py

```diff
@@ -0,0 +1,127 @@
+expected_dist = 37
+known_width = 4.18 ##inches
+focal_length = 1.37478492e+03
+from electrical.motor_controller import BasicMotorController
+import numpy as np
+import cv2
+import time 
+from enum import Enum
+
+arucoDict = cv2.aruco.Dictionary_get(cv2.aruco.DICT_APRILTAG_36h11)
+arucoParams = cv2.aruco.DetectorParameters_create()
+cap = cv2.VideoCapture(0)
+window = 'Camera'
+cv2.namedWindow(window)
+def distance_to_camera(knownWidth, focalLength, perWidth):
+  # compute and return the distance from the maker to the camera
+  return (knownWidth * focalLength) / perWidth
+
+def direction_of_tag(corner,image):
+  (_, _, _, _) = corner
+  _,w = image.shape[0], image.shape[1]
+  cx, _ = int(960),int(540)
+  if  bottomLeft[0] > cx:
+    return "right"
+  else:
+    return "left"
+
+idDict = {1: "R1", 2: "R2", 3: "R3"}#TODO add all the other tags
+visited = []
+timeout =  10
+angle_zero = 10
+step1_done = False
+step2_done = False
+step3_done = False 
+step4_done = False
+motor = BasicMotorController(True)
+start_time = None
+
+while True:
+    if start_time is None:
+        start_time = time.time()
+    ret, image = cap.read()
+    (corners, ids, rejected) = cv2.aruco.detectMarkers(image, arucoDict, parameters=arucoParams)
+    if len(corners) == 0:
+      if time.time() - start_time > timeout and step1_done is False:
+        #Step 1 - Check for april tags rn by rotating in circle for 15 seconds.
+          motor.turn_right()
+      if step1_done and step2_done is False:
+        #Step 2 - go back until len(corners) = 1 or when we see at least 1 april tag
+        motor.reverse()
+    if len(corners) > 0:
+      if step1_done:
+        step2_done = True
```

Comment:
this logic doesn't make sense to me. If step1 is done, step2 is done? What's the point of having step2 then?

Reply:
In step 1 we align with the center of the tag and move forward until the tag is out of sight and this all takes place under len(corners) > 0. Then we jump into step 2 which should start when we see no April tags(which is in the if step1_done and not step2_done condition which is under the if(Len(corners)<0) condition) . We reverse until we see an April tag which is when if len(corner) > 0. When an April tag is spotted, we set step2_done to true (we say we do this only if step1_done is true because we don't want to set step2_done to true where we just start our algorithm and we haven't even done step1.